File: backend/account-service/src/main/java/com/example/accountservice/service/BeneficiaryService.java
package com.example.accountservice.service;

import com.example.accountservice.dto.request.CreateBeneficiaryRequest;
import com.example.accountservice.dto.request.UpdateBeneficiaryRequest;
import com.example.accountservice.dto.response.BeneficiaryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Service for managing beneficiaries (danh bạ thụ hưởng)
 */
public interface BeneficiaryService {

    /**
     * Create a new beneficiary
     * @param customerId the customer ID who owns this beneficiary
     * @param request the beneficiary details
     * @return created beneficiary
     */
    BeneficiaryResponse createBeneficiary(String customerId, CreateBeneficiaryRequest request);

    /**
     * Update beneficiary details
     * @param customerId the customer ID
     * @param beneficiaryId the beneficiary ID
     * @param request the update details
     * @return updated beneficiary
     */
    BeneficiaryResponse updateBeneficiary(String customerId, String beneficiaryId, UpdateBeneficiaryRequest request);

    /**
     * Delete a beneficiary
     * @param customerId the customer ID
     * @param beneficiaryId the beneficiary ID
     */
    void deleteBeneficiary(String customerId, String beneficiaryId);

    /**
     * Get beneficiary by ID
     * @param customerId the customer ID
     * @param beneficiaryId the beneficiary ID
     * @return beneficiary details
     */
    BeneficiaryResponse getBeneficiary(String customerId, String beneficiaryId);

    /**
     * Get all beneficiaries for a customer
     * @param customerId the customer ID
     * @return list of beneficiaries
     */
    List<BeneficiaryResponse> getAllBeneficiaries(String customerId);

    /**
     * Get beneficiaries with pagination
     * @param customerId the customer ID
     * @param pageable pagination parameters
     * @return page of beneficiaries
     */
    Page<BeneficiaryResponse> getBeneficiaries(String customerId, Pageable pageable);

    /**
     * Search beneficiaries by name or nickname
     * @param customerId the customer ID
     * @param searchTerm the search term
     * @return list of matching beneficiaries
     */
    List<BeneficiaryResponse> searchBeneficiaries(String customerId, String searchTerm);

    /**
     * Get most frequently used beneficiaries
     * @param customerId the customer ID
     * @param limit maximum number of results
     * @return list of most used beneficiaries
     */
    List<BeneficiaryResponse> getMostUsedBeneficiaries(String customerId, int limit);

    /**
     * Update transfer statistics after a successful transfer
     * @param customerId the customer ID
     * @param beneficiaryAccountNumber the beneficiary account number
     */
    void updateTransferStats(String customerId, String beneficiaryAccountNumber);
}

File: frontend/fe-customer/app/service/beneficiaryService.ts
import { isAxiosError } from 'axios'
import axiosAccount from '~/config/axiosAccount'
import type { Beneficiary, CreateBeneficiaryRequest, UpdateBeneficiaryRequest } from '~/type/beneficiary'

const normalizeError = (error: unknown) => {
  if (isAxiosError(error)) {
    const message = (error.response?.data as { message?: string } | undefined)?.message
    return new Error(message ?? 'Có lỗi xảy ra, vui lòng thử lại')
  }
  return error instanceof Error ? error : new Error('Có lỗi xảy ra, vui lòng thử lại')
}

interface ApiResponse<T> {
  success: boolean
  message: string
  data: T
}

export const beneficiaryService = {
  async getAllBeneficiaries(customerId: string): Promise<Beneficiary[]> {
    try {
      const { data } = await axiosAccount.get<ApiResponse<Beneficiary[]>>(
        `/customers/${customerId}/beneficiaries`
      )
      
      if (!data?.success || !data?.data) {
        throw new Error(data?.message ?? 'Không thể lấy danh sách người thụ hưởng')
      }
      
      return data.data
    } catch (error) {
      throw normalizeError(error)
    }
  },

  async getBeneficiary(customerId: string, beneficiaryId: string): Promise<Beneficiary> {
    try {
      const { data } = await axiosAccount.get<ApiResponse<Beneficiary>>(
        `/customers/${customerId}/beneficiaries/${beneficiaryId}`
      )
      
      if (!data?.success || !data?.data) {
        throw new Error(data?.message ?? 'Không tìm thấy người thụ hưởng')
      }
      
      return data.data
    } catch (error) {
      throw normalizeError(error)
    }
  },

  async createBeneficiary(customerId: string, request: CreateBeneficiaryRequest): Promise<Beneficiary> {
    try {
      const { data } = await axiosAccount.post<ApiResponse<Beneficiary>>(
        `/customers/${customerId}/beneficiaries`,
        request
      )
      
      if (!data?.success || !data?.data) {
        throw new Error(data?.message ?? 'Không thể tạo người thụ hưởng')
      }
      
      return data.data
    } catch (error) {
      throw normalizeError(error)
    }
  },

  async updateBeneficiary(
    customerId: string, 
    beneficiaryId: string, 
    request: UpdateBeneficiaryRequest
  ): Promise<Beneficiary> {
    try {
      const { data } = await axiosAccount.put<ApiResponse<Beneficiary>>(
        `/customers/${customerId}/beneficiaries/${beneficiaryId}`,
        request
      )
      
      if (!data?.success || !data?.data) {
        throw new Error(data?.message ?? 'Không thể cập nhật người thụ hưởng')
      }
      
      return data.data
    } catch (error) {
      throw normalizeError(error)
    }
  },

  async deleteBeneficiary(customerId: string, beneficiaryId: string): Promise<void> {
    try {
      const { data } = await axiosAccount.delete<ApiResponse<void>>(
        `/customers/${customerId}/beneficiaries/${beneficiaryId}`
      )
      
      if (!data?.success) {
        throw new Error(data?.message ?? 'Không thể xóa người thụ hưởng')
      }
    } catch (error) {
      throw normalizeError(error)
    }
  },

  async searchBeneficiaries(customerId: string, searchTerm: string): Promise<Beneficiary[]> {
    try {
      const { data } = await axiosAccount.get<ApiResponse<Beneficiary[]>>(
        `/customers/${customerId}/beneficiaries/search?q=${encodeURIComponent(searchTerm)}`
      )
      
      if (!data?.success || !data?.data) {
        throw new Error(data?.message ?? 'Không thể tìm kiếm người thụ hưởng')
      }
      
      return data.data
    } catch (error) {
      throw normalizeError(error)
    }
  },

  async getMostUsedBeneficiaries(customerId: string, limit: number = 5): Promise<Beneficiary[]> {
    try {
      const { data } = await axiosAccount.get<ApiResponse<Beneficiary[]>>(
        `/customers/${customerId}/beneficiaries/most-used?limit=${limit}`
      )
      
      if (!data?.success || !data?.data) {
        throw new Error(data?.message ?? 'Không thể lấy danh sách thường dùng')
      }
      
      return data.data
    } catch (error) {
      throw normalizeError(error)
    }
  }
}
